servers/dsub/jobs/controllers/aggregation_controller.py

```diff
@@ -24,47 +20,109 @@ def get_job_aggregations(timeFrame, projectId=None):
+    provider = providers.get_provider(_provider_type(), projectId,
+                                      _auth_token())
+
+    jobs = jobs_generator.generate_jobs_by_window(provider, projectId,
+                                                  window_min)
+
+    total_summary = {}
+    user_summary = {}
+    job_name_summary = {}
+    label_summaries = {}
+
+    for job in jobs:
+        _count_total_summary(job, total_summary)
+        _count_for_key(job, user_summary, job.extensions.user_id)
+        _count_for_key(job, job_name_summary, job.name)
+        _count_top_labels(job, label_summaries)
+
+    aggregations = [
+        _to_aggregation('User Id', 'userId', user_summary),
+        _to_aggregation('Job Name', 'name', job_name_summary)
+    ] + (_to_top_labels_aggregations(label_summaries))
 
     return AggregationResponse(
-        summary=summaryCounts,
-        aggregations=[ownerAggregation, projectAggregation])
+        summary=_to_summary_counts(total_summary), aggregations=aggregations)
+
+
+def _count_total_summary(job, total_summary):
+    if job.status not in total_summary:
+        total_summary[job.status] = 0
+    total_summary[job.status] += 1
+
+
+def _count_for_key(job, summary, key):
+    count = summary.get(key, {})
+
+    if job.status not in count:
+        count[job.status] = 0
+    count[job.status] += 1
+
+    summary[key] = count
+
+
+def _count_top_labels(job, label_summaries):
+    for label in job.labels:
+        label_summary = label_summaries.get(label, {})
+        _count_for_key(job, label_summary, job.labels[label])
+        label_summaries[label] = label_summary
+
+
+def _to_summary_counts(summary_counts):
+    return StatusCounts([
+        StatusCount(status, count) for status, count in summary_counts.items()
+    ])
+
+
+def _to_aggregation(name, key, summary):
+    entries = []
+
+    for item, counts_dict in summary.items():
+        counts_list = []
+
+        for status, count in counts_dict.items():
+            counts_list.append(StatusCount(status=status, count=count))
+
+        entries.append(
+            AggregationEntry(
+                label=item, status_counts=StatusCounts(counts_list)))
+
+    return Aggregation(name=name, key=key, entries=entries)
+
+
+def _to_top_labels_aggregations(label_summaries):
+    # Rank the label summaries by the sum of valid item,
+    # where valid means a label has jobs more than _LABEL_VALID_MIN.
+    label_freq = {}
+    for label, item in label_summaries.items():
+        total_count = 0
+        for label_value, counts in item.items():
+            total_count += sum(
+                v for _, v in counts.items() if v > _LABEL_VALID_MIN)
+        label_freq[label] = total_count
+
+    aggregations = []
+
+    # To avoid messy look of dashboard, instead of returning all the aggregations by label,
+    # we only return the top _NUM_TOP_LABEL result.
+    num_label = min(len(label_freq), _NUM_TOP_LABEL)
+    for k, _ in sorted(
```

Comment:
Actually I just tried this locally, seems you can just do `for k in ...`

Reply:
`k in sorted()` dose not work. `sorted(label_freq.items(), key=lambda x: x[1], reverse=True)` gives a list of tuple. If we let `k` to be the tuple, we need to use k[0], which is not as clear as `k, _` in my opinion. And we might not use `label_freq.values()` since we need the key.

Reply:
OK, sounds good. I missed that this was a list of tuples rather than a dict.